File: BraintreeCore/src/test/java/com/braintreepayments/api/DeviceInspectorUnitTest.java
package com.braintreepayments.api;

import static com.braintreepayments.api.DeviceInspector.VENMO_BASE_64_ENCODED_SIGNATURE;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;

import org.json.JSONException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.util.ReflectionHelpers;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;

@RunWith(RobolectricTestRunner.class)
public class DeviceInspectorUnitTest {

    private Context context;
    private Configuration configuration;
    private ConnectivityManager connectivityManager;

    private PackageManager packageManager;

    private AppHelper appHelper;
    private ClassHelper classHelper;
    private UUIDHelper uuidHelper;
    private SignatureVerifier signatureVerifier;

    private Runtime runtime;
    private Process process;
    private File superUserApkFile;

    private DeviceInspector sut;

    @Before
    public void beforeEach() throws PackageManager.NameNotFoundException {
        context = mock(Context.class);
        Resources resources = mock(Resources.class);
        connectivityManager = mock(ConnectivityManager.class);
        configuration = new Configuration();

        packageManager = mock(PackageManager.class);

        appHelper = mock(AppHelper.class);
        uuidHelper = mock(UUIDHelper.class);
        classHelper = mock(ClassHelper.class);
        signatureVerifier = mock(SignatureVerifier.class);

        superUserApkFile = mock(File.class);
        runtime = mock(Runtime.class);
        process = mock(Process.class);

        when(context.getPackageManager()).thenReturn(packageManager);
        when(context.getPackageName()).thenReturn("com.sample.app");

        when(context.getResources()).thenReturn(resources);
        when(resources.getConfiguration()).thenReturn(configuration);

        when(context.getSystemService(Context.CONNECTIVITY_SERVICE)).thenReturn(connectivityManager);

        sut = new DeviceInspector(appHelper, classHelper, uuidHelper, signatureVerifier, runtime, superUserApkFile);
    }

    @Test
    public void getDeviceMetadata_returnsPlatform() throws JSONException {
        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("Android", metadata.toJSON().getString("platform"));
    }

    @Test
    public void getDeviceMetadata_returnsPlatformVersion() throws JSONException {
        ReflectionHelpers.setStaticField(Build.VERSION.class, "SDK_INT", 123);

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("123", metadata.toJSON().getString("platformVersion"));
    }

    @Test
    public void getDeviceMetadata_returnsSDKVersion() throws JSONException {
        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals(BuildConfig.VERSION_NAME, metadata.toJSON().getString("sdkVersion"));
    }

    @Test
    public void getDeviceMetadata_returnsMerchantAppId() throws JSONException {
        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("com.sample.app", metadata.toJSON().getString("merchantAppId"));
    }

    @Test
    public void getDeviceMetadata_returnsApplicationNameUnknownByDefault() throws JSONException {
        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("ApplicationNameUnknown", metadata.toJSON().getString("merchantAppName"));
    }

    @Test
    public void getDeviceMetadata_whenApplicationInfoUnavailable_returnsApplicationNameUnknown() throws PackageManager.NameNotFoundException, JSONException {
        when(packageManager.getApplicationInfo("com.sample.app", 0)).thenThrow(new PackageManager.NameNotFoundException());

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("ApplicationNameUnknown", metadata.toJSON().getString("merchantAppName"));
    }

    @Test
    public void getDeviceMetadata_returnsAppNameFromPackageManager() throws PackageManager.NameNotFoundException, JSONException {
        ApplicationInfo applicationInfo = new ApplicationInfo();
        when(packageManager.getApplicationInfo("com.sample.app", 0)).thenReturn(applicationInfo);
        when(packageManager.getApplicationLabel(applicationInfo)).thenReturn("SampleAppName");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("SampleAppName", metadata.toJSON().getString("merchantAppName"));
    }

    @Test
    public void getDeviceMetadata_whenApplicationInfoNotFound_returnsAppNameFromPackageManager() throws PackageManager.NameNotFoundException, JSONException {
        when(packageManager.getApplicationInfo("sample-package-name", 0)).thenThrow(new PackageManager.NameNotFoundException());

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("ApplicationNameUnknown", metadata.toJSON().getString("merchantAppName"));
    }

    @Test
    public void getDeviceMetadata_returnsDeviceManufacturer() throws JSONException {
        ReflectionHelpers.setStaticField(Build.class, "MANUFACTURER", "device-manufacturer");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("device-manufacturer", metadata.toJSON().getString("deviceManufacturer"));
    }

    @Test
    public void getDeviceMetadata_returnsDeviceModel() throws JSONException {
        ReflectionHelpers.setStaticField(Build.class, "MODEL", "device-model");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("device-model", metadata.toJSON().getString("deviceModel"));
    }

    @Test
    public void getDeviceMetadata_returnsDeviceAppGeneratedPersistentUuid() throws JSONException {
        when(uuidHelper.getPersistentUUID(context)).thenReturn("persistent-uuid");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("persistent-uuid",
                metadata.toJSON().getString("deviceAppGeneratedPersistentUuid"));
    }

    @Test
    public void getDeviceMetadata_whenBuildProductManufacturerAndFingerprintAreValid_returnsFalseForIsSimulator() throws JSONException {
        ReflectionHelpers.setStaticField(Build.class, "PRODUCT", "randomBuildProduct");
        ReflectionHelpers.setStaticField(Build.class, "MANUFACTURER", "randomBuildManufacturer");
        ReflectionHelpers.setStaticField(Build.class, "FINGERPRINT", "randomBuildFingerprint");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("false", metadata.toJSON().getString("isSimulator"));
    }

    @Test
    public void getDeviceMetadata_whenBuildProductIsGoogleSdk_returnsTrueForIsSimulator() throws JSONException {
        ReflectionHelpers.setStaticField(Build.class, "PRODUCT", "google_sdk");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertTrue(metadata.toJSON().getBoolean("isSimulator"));
    }

    @Test
    public void getDeviceMetadata_whenBuildProductIsSdk_returnsTrueForIsSimulator() throws JSONException {
        ReflectionHelpers.setStaticField(Build.class, "PRODUCT", "sdk");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertTrue(metadata.toJSON().getBoolean("isSimulator"));
    }

    @Test
    public void getDeviceMetadata_whenBuildManufacturerIsGenymotion_returnsTrueForIsSimulator() throws JSONException {
        ReflectionHelpers.setStaticField(Build.class, "MANUFACTURER", "Genymotion");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertTrue(metadata.toJSON().getBoolean("isSimulator"));
    }

    @Test
    public void getDeviceMetadata_whenBuildFingerpintContainsGeneric_returnsTrueForIsSimulator() throws JSONException {
        ReflectionHelpers.setStaticField(Build.class, "FINGERPRINT", "generic");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertTrue(metadata.toJSON().getBoolean("isSimulator"));
    }

    @Test
    public void getDeviceMetadata_forwardsSessionId() throws JSONException {
        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("session-id", metadata.toJSON().getString("sessionId"));
    }

    @Test
    public void getDeviceMetadata_forwardsIntegrationType() throws JSONException {
        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("integration-type", metadata.toJSON().getString("integrationType"));
    }

    @Test
    public void getDeviceMetadata_returnsNetworkType() throws JSONException {
        NetworkInfo networkInfo = mock(NetworkInfo.class);
        when(networkInfo.getTypeName()).thenReturn("network-type-name");
        when(connectivityManager.getActiveNetworkInfo()).thenReturn(networkInfo);

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("network-type-name", metadata.toJSON().getString("deviceNetworkType"));
    }

    @Test
    public void getDeviceMetadata_whenNetworkInfoUnavailable_returnsNoneForNetworkType() throws JSONException {
        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("none", metadata.toJSON().getString("deviceNetworkType"));
    }

    @Test
    public void getDeviceMetadata_whenDeviceOrientationIsPortrait_returnsPortraitForUserInterfaceOrientation() throws JSONException {
        configuration.orientation = Configuration.ORIENTATION_PORTRAIT;

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("Portrait", metadata.toJSON().getString("userInterfaceOrientation"));
    }

    @Test
    public void getDeviceMetadata_whenDeviceOrientationIsLandscape_returnsLandscapeForUserInterfaceOrientation() throws JSONException {
        configuration.orientation = Configuration.ORIENTATION_LANDSCAPE;

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("Landscape", metadata.toJSON().getString("userInterfaceOrientation"));
    }

    @Test
    public void getDeviceMetadata_whenDeviceOrientationIsUndefined_returnsUnknownForUserInterfaceOrientation() throws JSONException {
        configuration.orientation = Configuration.ORIENTATION_UNDEFINED;

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("Unknown", metadata.toJSON().getString("userInterfaceOrientation"));
    }

    @Test
    public void getDeviceMetadata_returnsAppVersion() throws JSONException, PackageManager.NameNotFoundException {
        PackageInfo packageInfo = new PackageInfo();
        packageInfo.versionName = "AppVersion";
        when(packageManager.getPackageInfo("com.sample.app", 0)).thenReturn(packageInfo);

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("AppVersion", metadata.toJSON().getString("merchantAppVersion"));
    }

    @Test
    public void getDeviceMetadata_whenAppVersionUnavailable_returnsVersionUnknownByDefault() throws JSONException, PackageManager.NameNotFoundException {
        when(packageManager.getPackageInfo("com.sample.app", 0)).thenThrow(new PackageManager.NameNotFoundException());

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("VersionUnknown", metadata.toJSON().getString("merchantAppVersion"));
    }

    @Test
    public void getDeviceMetadata_forwardsDropInVersionFromClassHelper() throws JSONException {
        String className = "com.braintreepayments.api.dropin.BuildConfig";
        String fieldName = "VERSION_NAME";
        when(classHelper.getFieldValue(className, fieldName)).thenReturn("drop-in-version");

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertEquals("drop-in-version", metadata.toJSON().getString("dropinVersion"));
    }

    @Test
    public void getDeviceMetadata_forwardsIsPayPalInstalledResultFromAppHelper() throws JSONException {
        when(appHelper.isAppInstalled(context, "com.paypal.android.p2pmobile")).thenReturn(true);

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertTrue(metadata.toJSON().getBoolean("paypalInstalled"));
    }

    @Test
    public void getDeviceMetadata_forwardsIsVenmoInstalledResultFromAppHelper() throws JSONException {
        when(appHelper.isAppInstalled(context, "com.venmo")).thenReturn(true);

        DeviceMetadata metadata = sut.getDeviceMetadata(context, "session-id", "integration-type");
        assertTrue(metadata.toJSON().getBoolean("venmoInstalled"));
    }

    @Test
    public void isPayPalInstalled_forwardsIsPayPalInstalledResultFromAppHelper() {
        when(appHelper.isAppInstalled(context, "com.paypal.android.p2pmobile")).thenReturn(true);
        assertTrue(sut.isPayPalInstalled(context));
    }

    @Test
    public void isVenmoInstalled_forwardsIsVenmoInstalledResultFromAppHelper() {
        when(appHelper.isAppInstalled(context, "com.venmo")).thenReturn(true);
        assertTrue(sut.isVenmoInstalled(context));
    }

    @Test
    public void isVenmoAppSwitchAvailable_checksForVenmoIntentAvailability() {
        sut.isVenmoAppSwitchAvailable(context);

        ArgumentCaptor<Intent> captor = ArgumentCaptor.forClass(Intent.class);
        verify(appHelper).isIntentAvailable(same(context), captor.capture());

        Intent intent = captor.getValue();
        ComponentName component = intent.getComponent();
        assertEquals("com.venmo", component.getPackageName());
        assertEquals("com.venmo.controller.SetupMerchantActivity", component.getClassName());
    }

    @Test
    public void isVenmoAppSwitchAvailable_whenVenmoIsNotInstalled_returnsFalse() {
        when(appHelper.isIntentAvailable(same(context), any(Intent.class))).thenReturn(false);
        assertFalse(sut.isVenmoAppSwitchAvailable(context));
    }

    @Test
    public void isVenmoAppSwitchAvailable_whenVenmoIsInstalledSignatureIsNotValid_returnsFalse() {
        when(appHelper.isIntentAvailable(same(context), any(Intent.class))).thenReturn(true);
        when(
                signatureVerifier.isSignatureValid(context, "com.venmo", VENMO_BASE_64_ENCODED_SIGNATURE)
        ).thenReturn(false);

        assertFalse(sut.isVenmoAppSwitchAvailable(context));
    }

    @Test
    public void isVenmoAppSwitchAvailable_whenVenmoIsInstalledSignatureIsValid_returnsTrue() {
        when(appHelper.isIntentAvailable(same(context), any(Intent.class))).thenReturn(true);
        when(
                signatureVerifier.isSignatureValid(context, "com.venmo", VENMO_BASE_64_ENCODED_SIGNATURE)
        ).thenReturn(true);

        assertTrue(sut.isVenmoAppSwitchAvailable(context));
    }
}

File: BraintreeCore/src/test/java/com/braintreepayments/api/DeviceInspectorUnitTest.kt
package com.braintreepayments.api

import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.content.res.Resources
import android.net.ConnectivityManager
import android.os.Build
import android.os.Build.VERSION
import io.mockk.*
import org.json.JSONException
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.util.ReflectionHelpers

@RunWith(RobolectricTestRunner::class)
class DeviceInspectorUnitTest {

    private var context: Context = mockk(relaxed = true)
    private var configuration: Configuration = mockk(relaxed = true)
    private var connectivityManager: ConnectivityManager = mockk(relaxed = true)
    private var packageManager: PackageManager = mockk(relaxed = true)
    private var appHelper: AppHelper = mockk(relaxed = true)
    private var signatureVerifier: SignatureVerifier = mockk(relaxed = true)
    private lateinit var sut: DeviceInspector

    private val btConfiguration =
        com.braintreepayments.api.Configuration.fromJson(Fixtures.CONFIGURATION_WITH_ENVIRONMENT)

    @Before
    @Throws(PackageManager.NameNotFoundException::class)
    fun beforeEach() {
        val resources = mockk<Resources>(relaxed = true)
        every { context.packageManager } returns packageManager
        every { context.packageName } returns "com.sample.app"
        every { context.getSystemService(Context.CONNECTIVITY_SERVICE) } returns connectivityManager
        every { context.resources } returns resources
        every { resources.configuration } returns configuration

        sut = DeviceInspector(
            appHelper,
            signatureVerifier,
        )
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_returnsPlatform() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("Android", metadata.toJSON().getString("platform"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_returnsEventSource() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("mobile-native", metadata.toJSON().getString("event_source"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_returnsComponent() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("braintreeclientsdk", metadata.toJSON().getString("comp"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_returnsPlatformVersion() {
        ReflectionHelpers.setStaticField(VERSION::class.java, "SDK_INT", 123)
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("Android API 123", metadata.toJSON().getString("client_os"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_returnsSDKVersion() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals(BuildConfig.VERSION_NAME, metadata.toJSON().getString("c_sdk_ver"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_returnsAppId() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("com.sample.app", metadata.toJSON().getString("app_id"))
    }

    @Test
    @Throws(PackageManager.NameNotFoundException::class, JSONException::class)
    fun getDeviceMetadata_whenApplicationInfoUnavailable_returnsApplicationNameUnknown() {
        every { packageManager.getApplicationInfo("com.sample.app", 0) } throws PackageManager.NameNotFoundException()
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals(
            "ApplicationNameUnknown",
            metadata.toJSON().getString("app_name")
        )
    }

    @Test
    @Throws(PackageManager.NameNotFoundException::class, JSONException::class)
    fun getDeviceMetadata_returnsAppNameFromPackageManager() {
        val applicationInfo = ApplicationInfo()
        every { packageManager.getApplicationInfo("com.sample.app", 0) } returns applicationInfo
        every { packageManager.getApplicationLabel(applicationInfo) } returns "SampleAppName"
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("SampleAppName", metadata.toJSON().getString("app_name"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_returnsDeviceManufacturer() {
        ReflectionHelpers.setStaticField(Build::class.java, "MANUFACTURER", "device-manufacturer")
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals(
            "device-manufacturer",
            metadata.toJSON().getString("device_manufacturer")
        )
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_returnsDeviceModel() {
        ReflectionHelpers.setStaticField(Build::class.java, "MODEL", "device-model")
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("device-model", metadata.toJSON().getString("mobile_device_model"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_whenBuildProductManufacturerAndFingerprintAreValid_returnsFalseForIsSimulator() {
        ReflectionHelpers.setStaticField(Build::class.java, "PRODUCT", "randomBuildProduct")
        ReflectionHelpers.setStaticField(
            Build::class.java,
            "MANUFACTURER",
            "randomBuildManufacturer"
        )
        ReflectionHelpers.setStaticField(Build::class.java, "FINGERPRINT", "randomBuildFingerprint")
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("false", metadata.toJSON().getString("is_simulator"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_whenBuildProductIsGoogleSdk_returnsTrueForIsSimulator() {
        ReflectionHelpers.setStaticField(Build::class.java, "PRODUCT", "google_sdk")
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertTrue(metadata.toJSON().getBoolean("is_simulator"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_whenBuildProductIsSdk_returnsTrueForIsSimulator() {
        ReflectionHelpers.setStaticField(Build::class.java, "PRODUCT", "sdk")
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertTrue(metadata.toJSON().getBoolean("is_simulator"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_whenBuildManufacturerIsGenymotion_returnsTrueForIsSimulator() {
        ReflectionHelpers.setStaticField(Build::class.java, "MANUFACTURER", "Genymotion")
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertTrue(metadata.toJSON().getBoolean("is_simulator"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_whenBuildFingerpintContainsGeneric_returnsTrueForIsSimulator() {
        ReflectionHelpers.setStaticField(Build::class.java, "FINGERPRINT", "generic")
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertTrue(metadata.toJSON().getBoolean("is_simulator"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_forwardsSessionId() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("session-id", metadata.toJSON().getString("session_id"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_forwardsIntegrationType() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("integration-type", metadata.toJSON().getString("api_integration_type"))
    }

    @Test
    @Throws(JSONException::class, PackageManager.NameNotFoundException::class)
    fun getDeviceMetadata_returnsAppVersion() {
        val packageInfo = PackageInfo()
        packageInfo.versionName = "AppVersion"
        every { packageManager.getPackageInfo("com.sample.app", 0) } returns packageInfo
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("AppVersion", metadata.toJSON().getString("mapv"))
    }

    @Test
    @Throws(JSONException::class, PackageManager.NameNotFoundException::class)
    fun getDeviceMetadata_whenAppVersionUnavailable_returnsVersionUnknownByDefault() {
        every { packageManager.getPackageInfo("com.sample.app", 0) } throws PackageManager.NameNotFoundException()
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("VersionUnknown", metadata.toJSON().getString("mapv"))
    }

    @Test
    @Throws(JSONException::class)
    fun getDeviceMetadata_forwardsDropInVersionFromClassHelper() {
        mockkObject(DeviceInspector) {
            every { DeviceInspector.getDropInVersion() } returns "fake-drop-in-version"
            val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
            assertEquals("fake-drop-in-version", metadata.toJSON().getString("drop_in_sdk_ver"))
        }
    }

    @Test
    @Throws(JSONException::class)
    fun getSDKEnvironment_forwardsEnvironmentFromConfig() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("test", metadata.toJSON().getString("merchant_sdk_env"))
    }

    @Test
    @Throws(JSONException::class)
    fun getMerchantID_forwardsMerchantIdFromConfig() {
        val metadata = sut.getDeviceMetadata(context, btConfiguration, "session-id", "integration-type")
        assertEquals("integration_merchant_id", metadata.toJSON().getString("merchant_id"))
    }

    @Test
    fun isPayPalInstalled_forwardsIsPayPalInstalledResultFromAppHelper() {
        every { appHelper.isAppInstalled(context, "com.paypal.android.p2pmobile") } returns true
        assertTrue(sut.isPayPalInstalled(context))
    }

    @Test
    fun isVenmoInstalled_forwardsIsVenmoInstalledResultFromAppHelper() {
        every { appHelper.isAppInstalled(context, "com.venmo") } returns true
        assertTrue(sut.isVenmoInstalled(context))
    }

    @Test
    fun isVenmoAppSwitchAvailable_checksForVenmoIntentAvailability() {
        sut.isVenmoAppSwitchAvailable(context)
        val intentSlot = slot<Intent>()
        verify { appHelper.isIntentAvailable(context, capture(intentSlot)) }

        val intent = intentSlot.captured
        val component = intent.component
        assertEquals("com.venmo", component!!.packageName)
        assertEquals("com.venmo.controller.SetupMerchantActivity", component.className)
    }

    @Test
    fun isVenmoAppSwitchAvailable_whenVenmoIsNotInstalled_returnsFalse() {
        every { appHelper.isIntentAvailable(context, ofType(Intent::class)) } returns false
        assertFalse(sut.isVenmoAppSwitchAvailable(context))
    }

    @Test
    fun isVenmoAppSwitchAvailable_whenVenmoIsInstalledSignatureIsNotValid_returnsFalse() {
        every { appHelper.isIntentAvailable(context, ofType(Intent::class)) } returns true

        every {
            signatureVerifier.isSignatureValid(
                context,
                "com.venmo",
                DeviceInspector.VENMO_BASE_64_ENCODED_SIGNATURE
            )
        } returns false

        assertFalse(sut.isVenmoAppSwitchAvailable(context))
    }

    @Test
    fun isVenmoAppSwitchAvailable_whenVenmoIsInstalledSignatureIsValid_returnsTrue() {
        every { appHelper.isIntentAvailable(context, ofType(Intent::class)) } returns true

        every {
            signatureVerifier.isSignatureValid(
                context,
                "com.venmo",
                DeviceInspector.VENMO_BASE_64_ENCODED_SIGNATURE
            )
        } returns true

        assertTrue(sut.isVenmoAppSwitchAvailable(context))
    }
}
